src/evict/evict_lru.c

```diff
@@ -1506,16 +1543,11 @@ __evict_walk(WT_SESSION_IMPL *session, WT_EVICT_QUEUE *queue)
              * have a saved handle, pick one randomly from the list.
              */
             if ((dhandle = cache->walk_tree) != NULL)
-                cache->walk_tree = NULL;
+                __clear_saved_walk_tree(session);
             else
                 __evict_walk_choose_dhandle(session, &dhandle);
         } else {
-            if (incr) {
-                WT_ASSERT(session, __wt_atomic_loadi32(&dhandle->session_inuse) > 0);
-                (void)__wt_atomic_subi32(&dhandle->session_inuse, 1);
-                incr = false;
-                cache->walk_tree = NULL;
-            }
+            __clear_saved_walk_tree(session);
```

Comment:
Let's assume that this is the second iteration in this for loop, and in the first iteration we set the pointer `dhandle`, however, we skip the `dhandle` selection for eviction (in which case we never incr `dhandle->session_inuse`). However, when calling `__clear_saved_walk_tree`. We have an assertion - 
```
WT_ASSERT(session, __wt_atomic_loadi32(&dhandle->session_inuse) > 0);
```

Reply:
Functions `__clear_saved_walk_tree` and `__set_saved_walk_tree` implement a "smart pointer" pattern.
If `cache->walk_tree` is not null, then it carries a use count.
Whenever it's set to a non-null value, the use count is incremented. Whenever a non-null value is cleared, the use count is decremented.

In this particular case, `__clear_saved_walk_tree` will clear `cache->walk_tree` and will decrement the use count iff it was not null.

Assignment is happening below via `__set_saved_walk_tree`.

Reply:
thanks! Makes sense! :-)